controllers/release/adapter.go

```diff
@@ -203,22 +205,35 @@ func (a *adapter) EnsureReleaseIsProcessed() (controller.OperationResult, error)
 			return controller.RequeueWithError(err)
 		}
 
+		if len(resources.ReleasePlan.Spec.Target) == 0 && resources.ReleasePlan.Spec.Pipeline == nil {
+			return controller.RequeueWithError(fmt.Errorf("tenant PipelineRun must have ReleasePlan.Spec.Pipeline defined"))
+		}
+
 		if pipelineRun == nil {
-			// Only create a RoleBinding if a ServiceAccount is specified
-			if roleBinding == nil && resources.ReleasePlanAdmission.Spec.Pipeline.ServiceAccount != "" {
-				// This string should probably be a constant somewhere
-				roleBinding, err = a.createRoleBindingForClusterRole("release-pipeline-resource-role", resources.ReleasePlanAdmission)
-				if err != nil {
-					return controller.RequeueWithError(err)
+			if resources.ReleasePlanAdmission != nil {
+				// Only create a RoleBinding if a ServiceAccount is specified
+				if roleBinding == nil && resources.ReleasePlanAdmission.Spec.Pipeline.ServiceAccount != "" {
+					// This string should probably be a constant somewhere
+					roleBinding, err = a.createRoleBindingForClusterRole("release-pipeline-resource-role", resources.ReleasePlanAdmission)
+					if err != nil {
+						return controller.RequeueWithError(err)
+					}
 				}
 			}
 
-			pipelineRun, err = a.createManagedPipelineRun(resources)
+			if len(resources.ReleasePlan.Spec.Target) == 0 {
```

Comment:
I think a better check here is if ReleasePlan has a pipeline in it

Reply:
I was actually doing this at first. I switched to the target because it would not make sense having a target for a tenant pipeline, and I ended up using this check as default, but if you think it is more readable using `pipeline` instead of target, also fine.

In my understanding, it only makes sense having target for set for managed pipelines. Could you comment on this statement? Maybe this is a  wrong assumption.

Reply:
I am not 100% on this, but TMBK this is how things are going to go:
- We add support for tenant pipelines (this PR). You can run either a managed pipeline or a tenant pipeline in your release, but not both
- We allow users to run both. A collector pipeline is pretty much just a specific tenant pipeline

So I don't want to pollute this PR too much with things for supporting both, but for things like this where the two approaches are just about the same, might as well go with the future proof one IMO

Reply:
I did not get which operation? the `EnsureReleaseIsProcessed`?  I did not get if you want this change or not.

Reply:
If it was to me, I would have implemented it in two different operations. That way they are shorter, easier to maintain and more idempotent. Letting @johnbieren decide here whether we go in that direction now, in the future or never ;)

Reply:
I think the suggestion is instead of modifying EnsureReleaseIsProcessed to support both managed and tenant pipelines, to create a new operation like EnsureTenantPipelinesAreExecuted or something to run the tenant pipeline and leave the managed operation alone (besides perhaps a check to skip it if a tenant pipeline is defined). I do think this is a bit simpler and will pay off as we add more to it in the future

Reply:
Please take a look now if the makes sense. I added the new Handler.